Add tests to stock details page.

BaggioWongHK opened the pull request:
Add unit tests for stock details page. 
Lint fixes and remove `.and.callThrough()` for spies.

Commit:
Add unit tests for stock details page.

Also included minor fixes for stock details page based on unit tests.

Commit:
Finish unit tests for stock details page.

Commit:
Lint fixes and remove callThrough().

Better to avoid really calling functions as much as possible.

nikhilmetrani commented on src/app/containers/stock-details-page/stock-details-page.component.spec.ts:
```diff
@@ -29,4 +33,139 @@ describe('StockDetailsPageComponent', () => {
+    });
+
+    it('should return true if stock exists', () => {
+      expect(component['stockExists']('aapl')).toEqual(true);
```
Any specific reason for using [] operator instead of .?
```
component['stockExists']('aapl')
```
instead of
```
component.stockExists('aapl')
```

Rest of the code base is using (.) operator and changing the style leads code style inconsistency.

BaggioWongHK replied:
This is the only way to access `private` properties and methods, (you can't do that with the dot operator `.`) - I thought about changing `private` properties to `public`, but they're only used privately so, I think the access modifier is appropriate.

BaggioWongHK replied:
Or rather, the issue is that it keeps on throwing a syntax error if I use the dot operator - I can change everything to `[]` instead.

nikhilmetrani replied:
It's better to make the two methods public in this case. Eventually, the methods will be encapsulated into a service for reusability.

BaggioWongHK replied:
Oh I see, I'll do that then, thanks.

nikhilmetrani commented on src/app/containers/stock-details-page/stock-details-page.component.spec.ts:
```diff
@@ -29,4 +33,139 @@ describe('StockDetailsPageComponent', () => {
+
+    it('should return false if stock doesn\'t exist', () => {
+      expect(component['stockExists']('A')).toEqual(false);
+      expect(component['stockExists']('')).toEqual(false);
```
It's best to limit one expect per it block. Multiple expects makes it hard to debug test failures as the codebase grows.

BaggioWongHK replied:
Oh I see, I can split up expect lines. :)

nikhilmetrani commented on src/app/containers/stock-details-page/stock-details-page.component.spec.ts:
```diff
@@ -29,4 +33,139 @@ describe('StockDetailsPageComponent', () => {
+    it('should buy units of a stock if units valid', () => {
+      component.buy('10000');
+      expect(storeSpy).toHaveBeenCalled();
+      component.buy('10000.01');
```
I don't think stock units can be bought in fractions. It should be whole numbers only.

BaggioWongHK replied:
Yes, but yesterday, remember we got rid of the custom method to check for a valid string and used `parseInt()` and `Number.isFinite()`? It turns out that parseInt('10000.01') truncates the digits after the decimal point. Should I bring back the regex check for situations like these?

nikhilmetrani replied:
Business logic should handle this by converting the number to whole number - Example - 
```
Math.floor(parseInt(stringValue));
```

BaggioWongHK replied:
I thought the issue was with the input being invalid in the first place? Shouldn't this be rejected instead of rounding it?

BaggioWongHK replied:
Hmm, I think we can just implement it like this for now, as per your suggestion.

nikhilmetrani replied:
In this particular case, the user should not be allowed to type anything apart from a whole number, however, there should be a secondary check in case someone manages to call the function directly.

BaggioWongHK replied:
I didn't think of that, I think this is even a better suggestion. I'll try to implement it.

nikhilmetrani commented on src/app/containers/stock-details-page/stock-details-page.component.spec.ts:
```diff
@@ -29,4 +33,139 @@ describe('StockDetailsPageComponent', () => {
+    it('should return false if there are not enough units of a stock', () => {
+      component['store'].select('portfolio').subscribe((portfolio) => {
+        const id = portfolio.stockItems[0].id;
+        expect(component['enoughStock']('300', 10)).not.toBeTruthy();
```
It's better to use toBeFalsy() instead of not.toBeTruthy().

BaggioWongHK replied:
Oh I didn't know there was a `toBeFalsy()` method, thanks for the heads up. :)

nikhilmetrani reviewed (commented):
Try to get line 60 and 129 covered -> [Report](https://104-132695385-gh.circle-artifacts.com/0/home/circleci/rangle-investing/coverage/src/app/containers/stock-details-page/stock-details-page.component.ts.html)

Commit:
Remove private modifiers.

Makes accessing private and public properties inconsistent in spec file.

Commit:
Remove all properties accessed through bracket notation.

After removing private modifiers, can now access properties through dot operator.
Privately injected variables can be retrieved fixture.debugElement.injector, also changed these to dot operator access.

Commit:
Separate out expectations into it blocks.

As per Nikhil's suggestion, multiple expect statements within an it block have now been separated.

Commit:
Add tests to increase test coverage.

Abstracted Array.prototype.find(callback) callback into a separate function to increase test coverage.
Added test to for the callback.
Added mock value for router param tickerSymbol.

Commit:
Guards against invalid input in buy form.

Removed invalid import (Observable.of to of) in spec file.
Added buyInputValidator to prevent users from entering non digit values
(deletes invalid values in form automatically).
Added test for buyInputValidator.
Added testInteger method to prevent users from entering non digit values
(this is for an extra layer of check internally in the buy and sell methods).
Added tests for testInteger method.
Fixed bug where the stockDetails page is unreachable because the stocks
array was not populated before it was used in the check condition.
Added checks in buy and sell methods to prevent decimal numbers from being
entered.

Commit:
Change .not.toBeTruthy to toBeFalsy.

BaggioWongHK commented:
@nikhilmetrani I've implemented all the changes you've recommended, please take a look if you have time! :)
